plutus-ir/src/Language/PlutusIR/Compiler/Provenance.hs

```diff
@@ -25,8 +25,19 @@ data Provenance a = Original a
                   | TypeBinding String (Provenance a)
                   | DatatypeComponent DatatypeComponent (Provenance a)
                   | NoProvenance
+                  | MultipleSources [Provenance a]
                   deriving (Show, Eq)
 
+-- Needed for LetFloat transformation to merge annotations
+instance Semigroup (Provenance a) where
+  MultipleSources ps1 <> MultipleSources ps2 = MultipleSources (ps1++ps2)
+  x <> MultipleSources ps2 = MultipleSources (x:ps2)
+  MultipleSources ps1 <> x = MultipleSources (ps1++[x])
+  x <> y = MultipleSources [x,y]
```

Comment:
That's a very funny `Semigroup` instance, never seen anything of this kind.

Maybe make it `MultipleSources (DList (Provenance a))`? Or I guess this is never going to be a bottleneck and we're fine with the inefficient appends?

Reply:
I removed NoProvenance data constructor and replaced it to mempty == = MultipleSources []  . This is still not perfect. I tried @effectfully suggestion to  use [Provenance a] instead of Provenance a, but it needed way many changes across the pir compiler and plutus-core. So i left it for later.